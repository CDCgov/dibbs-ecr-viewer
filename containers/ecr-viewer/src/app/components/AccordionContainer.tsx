import React from "react";
import { Accordion, HeadingLevel } from "@trussworks/react-uswds";
import { toKebabCase } from "@/app/utils/format-utils";
import { RenderableNode } from "@/app/utils/data-utils";

export interface AccordionItemProps {
  title: RenderableNode;
  content: React.ReactNode;
  expanded: boolean;
  id: string;
  className?: string;
  headingLevel: HeadingLevel;
  handleToggle?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

type AccordionContainerProps = {
  accordionItems: AccordionItemProps[];
};

/**
 * Functional component for an accordion container displaying various sections of eCR information.
 * @param props - Props containing FHIR bundle and path mappings.
 * @param props.accordionItems - The list of accordion items.
 * @returns The JSX element representing the accordion container.
 */
const AccordionContainer = ({
  accordionItems,
}: AccordionContainerProps): React.JSX.Element => {
  const items: AccordionItemProps[] = accordionItems.map((item, index) => {
    let formattedTitle = toKebabCase(`${item["title"]}`);
    return {
      ...item,
      id: `${formattedTitle}_${index + 1}`,
      title: <span id={formattedTitle}>{item["title"]}</span>,
    };
  });

  return (
    <Accordion
      className="info-container"
      items={items}
      multiselectable={true}
    />
  );
};
export default AccordionContainer;
