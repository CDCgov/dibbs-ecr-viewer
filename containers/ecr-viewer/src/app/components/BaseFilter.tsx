import React, {
  ComponentType,
  ReactNode,
  useCallback,
  useContext,
  useEffect,
  useRef,
} from "react";
import { Button, Label } from "@trussworks/react-uswds";
import { useSearchParams } from "next/navigation";
import {
  FILTER_CLOSED,
  FILTER_SUBMITTED,
  FilterOpenContext,
} from "@/app/components/Filters";
import { toKebabCase } from "@/app/utils/format-utils";

/**
 * A reusable Filter component for eCR Library. It displays a button
 * that toggles a dropdown form for filtering data. The form includes functionality
 * for resetting and submitting the filter.
 * @param props - The props for the Filter component.
 * @param props.isActive - Boolean to indicate if a filter is actively filtering eCRs.
 * @param props.type - Type of the filter (e.g., "Recieved Date", "Reportable Condition").
 * @param props.title - Title text displayed on the button; defaults to `type`.
 * @param props.icon - Icon component rendered inside the filter button.
 * @param props.tag - Optional tag element displayed next to the title.
 * @param props.resetHandler - Callback for resetting the filter.
 * @param props.submitHandler - Callback for applying the filter on form submission.
 * @param props.children - The filter form fields and content displayed in the dropdown.
 * @returns A JSX element for the filter with a dropdown form.
 */
export const Filter = ({
  isActive,
  type,
  title = "",
  icon: IconTag,
  tag = "",
  resetHandler,
  submitHandler,
  children,
}: {
  isActive: boolean;
  type: string;
  title?: string;
  icon: ComponentType<{ className?: string }>;
  tag?: ReactNode;
  resetHandler: () => void;
  submitHandler: () => void;
  children: ReactNode;
}) => {
  const { filterBoxOpen, setFilterBoxOpen, lastOpenButtonRef } =
    useContext(FilterOpenContext);
  const openBtnRef = useRef<HTMLElement | null>(null);
  const searchParams = useSearchParams();

  const isFilterBoxOpen = filterBoxOpen === type;
  const setIsFilterBoxOpen = useCallback((open: boolean) => {
    if (open) {
      setFilterBoxOpen(type);
      // Set the last open button to this button when we open it
      lastOpenButtonRef.current = openBtnRef.current?.parentElement || null;
    } else {
      setFilterBoxOpen(FILTER_CLOSED);
    }
    openBtnRef?.current?.parentElement?.focus();
  }, []);

  // This filter has closed. We need the special submitted case to prevent
  // a race condition with submitting and resetting if we try to do a reset
  // just after submitting.
  useEffect(() => {
    if (filterBoxOpen !== FILTER_SUBMITTED && filterBoxOpen !== type) {
      resetHandler();
    }
  }, [filterBoxOpen]);

  // Force a reset if the search params update
  useEffect(resetHandler, [searchParams]);

  return (
    <div onClick={(e) => e.stopPropagation()}>
      <div className="position-relative display-flex flex-column">
        <Button
          className={`margin-right-0 ${
            isActive ? "filters-applied" : "filter-button"
          }`}
          aria-label={`Filter by ${type}`}
          aria-haspopup="listbox"
          aria-expanded={isFilterBoxOpen}
          onClick={() => {
            setIsFilterBoxOpen(!isFilterBoxOpen);
          }}
          type="button"
        >
          <span ref={openBtnRef} className="square-205 usa-icon">
            <IconTag aria-hidden className="square-205" />
          </span>
          <span className="text-ink">{title || type}</span>
          {tag && (
            <span
              className="usa-tag padding-05 bg-base-darker radius-md"
              data-testid="filter-tag"
            >
              {tag}
            </span>
          )}
        </Button>

        {isFilterBoxOpen && (
          <div className="usa-combo-box top-full left-0">
            <form
              onSubmit={(e) => {
                e.preventDefault();
                submitHandler();
                setFilterBoxOpen(FILTER_SUBMITTED);
                openBtnRef?.current?.parentElement?.focus();
              }}
            >
              <fieldset className="usa-combo-box border-0 padding-0 margin-top-1 bg-white position-absolute radius-md shadow-2 z-top maxh-6205 width-full">
                <FilterLegend type={type} />
                {children}
                <ApplyFilterButton type={type} />
              </fieldset>
            </form>
          </div>
        )}
      </div>
    </div>
  );
};

/**
 * A component to render a filter legend (title).
 * @param props - React props
 * @param props.type - The type of filter
 * @returns - The rendered legend element
 */
const FilterLegend = ({ type }: { type: string }) => {
  return (
    <legend className="line-height-sans-6 text-bold font-sans-xs bg-white width-full padding-y-1 padding-x-105 text-no-wrap">
      Filter by {type}
    </legend>
  );
};

/**
 * A button component for applying a filter.
 * @param props - React props
 * @param props.type - The type of filter
 * @returns - The rendered button element
 */
const ApplyFilterButton = ({ type }: { type: string }) => {
  return (
    <div className="display-flex flex-column flex-stretch padding-x-105">
      <Button
        type="submit"
        className="margin-y-1 margin-x-0 padding-y-1 padding-x-205 flex-fill text-no-wrap"
        aria-label={`Apply Filter for ${type}`}
      >
        Apply Filter
      </Button>
    </div>
  );
};

/**
 * A reusable radio button component, used for the filter by date feature.
 * @param props - The properties for the RadioDateOption component.
 * @param props.groupName - The name of the radio button group.
 * @param props.option - The value of the radio option.
 * @param props.label - The label to display next to the radio button.
 * @param props.onChange - The callback function to handle the `onChange` event when the radio button is clicked.
 * @param props.isChecked - Determines if the radio button is selected based on the current state.
 * @param props.classNames - (Optional) Additional custom CSS class names to apply to the radio button wrapper.
 * @returns The rendered RadioDateOption component.
 */
export const RadioDateOption = ({
  groupName,
  option,
  label,
  onChange,
  isChecked,
  classNames,
}: {
  groupName: string;
  option: string;
  label: string;
  onChange: (value: string) => void;
  isChecked: boolean;
  classNames?: string;
}) => {
  return (
    <div className={`checkbox-color usa-radio padding-x-105 ${classNames}`}>
      <input
        id={`${groupName}-${option}`}
        className="usa-radio__input"
        type="radio"
        name={`${groupName}-options`}
        value={option}
        onChange={(e) => onChange(e.target.value)}
        checked={isChecked}
      />
      <label
        className="line-height-sans-6 font-sans-xs margin-y-0 usa-radio__label text-no-wrap"
        htmlFor={`${groupName}-${option}`}
      >
        {label}
      </label>
    </div>
  );
};

/**
 * A group of radio button components, given a set of options.
 * @param props - The properties for the RadioDateOption component.
 * @param props.groupName - The name of the radio button group.
 * @param props.optionsMap - A map with each option as the key, and the corresponding labels as the value.
 * @param props.onChange - The callback function to handle the `onChange` event when the radio button is clicked.
 * @param props.currentOption - The option currently selected.
 * @param props.classNames - (Optional) Additional custom CSS class names to apply to the radio button wrapper.
 * @returns The rendered RadioDateOption component.
 */
export const RadioDateOptions = ({
  groupName,
  optionsMap,
  onChange,
  currentOption,
  classNames,
}: {
  groupName: string;
  optionsMap: Record<string, string>;
  onChange: (value: string) => void;
  currentOption: string;
  classNames?: string;
}) => {
  return (
    <>
      {Object.entries(optionsMap).map(([option, label]) => (
        <RadioDateOption
          key={`${groupName}-${option}`}
          groupName={groupName}
          option={option}
          label={label}
          onChange={onChange}
          isChecked={currentOption === option}
          classNames={classNames}
        />
      ))}
    </>
  );
};

/**
 *  A custom date input component for selecting a date.
 * @param props - The properties for the CustomDateInput component.
 * @param props.label - The label of the custom date input component.
 * @param props.onDateChange - The function that is caulled when the date changes.
 * @param props.defaultValue - The default value of the date input.
 * @param props.isRequired - Boolean indicating whether or not the date is required.
 * @param props.minValue - The minimum value of the date input.
 * @returns A JSX element containing a date input field and corresponding label.
 */
export const CustomDateInput = ({
  label,
  onDateChange,
  defaultValue,
  isRequired,
  minValue,
}: {
  label: string;
  onDateChange: (date: string) => void;
  defaultValue: string;
  isRequired: boolean;
  minValue?: string;
}) => {
  const today = new Date().toLocaleDateString("en-CA");
  const id = toKebabCase(label);
  return (
    <div>
      <Label htmlFor={id} className="margin-top-1">
        {label}
      </Label>
      <input
        id={id}
        data-testid={id}
        type="date"
        className="usa-input width-card margin-top-0 border-base-dark"
        defaultValue={defaultValue}
        min={minValue}
        max={today}
        required={isRequired}
        aria-label={label}
        onChange={(e) => {
          const date = e.target.value;
          onDateChange(date);
        }}
      />
    </div>
  );
};
