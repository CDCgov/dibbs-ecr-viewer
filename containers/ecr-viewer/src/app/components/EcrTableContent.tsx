import React from "react";

import { listEcrData } from "@/app/services/listEcrDataService";
import { DateRangePeriod } from "@/app/utils/date-utils";

import { EcrTableDataRow } from "./EcrTableDataRow";

/**
 * eCR Table
 * @param props - The properties passed to the component.
 * @param props.currentPage - The current page to be displayed
 * @param props.itemsPerPage - The number of items to be displayed in the table
 * @param props.sortColumn - The column to sort by
 * @param props.sortDirection - The direction to sort by
 * @param props.filterDates - The date range used to filter data
 * @param props.searchTerm - The search term used to list data
 * @param props.filterConditions - (Optional) The reportable condition(s) used to filter the data
 * @returns - eCR Table element
 */
const EcrTableContent = async ({
  currentPage,
  itemsPerPage,
  sortColumn,
  sortDirection,
  searchTerm,
  filterConditions,
  filterDates,
}: {
  currentPage: number;
  itemsPerPage: number;
  sortColumn: string;
  sortDirection: string;
  filterDates: DateRangePeriod;
  searchTerm?: string;
  filterConditions?: string[];
}) => {
  const startIndex = (currentPage - 1) * itemsPerPage;

  const data = await listEcrData(
    startIndex,
    itemsPerPage,
    sortColumn,
    sortDirection,
    filterDates,
    searchTerm,
    filterConditions,
  );

  return (
    <tbody>
      {data.map((item, index) => (
        <EcrTableDataRow key={index} item={item} />
      ))}
    </tbody>
  );
};

export default EcrTableContent;
