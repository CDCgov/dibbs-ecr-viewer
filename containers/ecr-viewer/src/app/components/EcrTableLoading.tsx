import { range } from "../utils/data-utils";

/**
 * The Ecr Library table, but with blobs instead of data.
 * @returns - The JSX element representing the eCR table.
 */
export const EcrTableLoading = () => {
  return (
    <tbody>
      {range(10).map((i) => {
        return (
          <BlobRow key={i} themeColor={i % 2 === 0 ? "gray" : "dark-gray"} />
        );
      })}
    </tbody>
  );
};

const Blob = ({ themeColor }: { themeColor: string }) => {
  return (
    <div className="grid-row">
      <div
        className={`loading-blob grid-col-8 loading-blob-${themeColor} width-full`}
      >
        &nbsp;
      </div>
    </div>
  );
};

const BlobRow = ({ themeColor }: { themeColor: string }) => {
  return (
    <tr>
      <td>
        <Blob themeColor={themeColor} />
      </td>
      <td>
        <Blob themeColor={themeColor} />
        <br />
        <Blob themeColor={themeColor} />
      </td>
      <td>
        <Blob themeColor={themeColor} />
        <br />
        <Blob themeColor={themeColor} />
      </td>
      <td>
        <Blob themeColor={themeColor} />
      </td>
      <td>
        <Blob themeColor={themeColor} />
      </td>
    </tr>
  );
};
