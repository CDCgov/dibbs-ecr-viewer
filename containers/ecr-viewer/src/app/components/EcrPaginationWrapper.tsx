"use client";

import { Label, Select } from "@trussworks/react-uswds";
import React, { ReactNode, useEffect, useState } from "react";
import { Pagination } from "@/app/components/Pagination";
import { usePathname, useRouter, useSearchParams } from "next/navigation";

interface EcrPaginationWrapperProps {
  totalCount: number;
  children: ReactNode;
}

interface UserPreferences {
  itemsPerPage: number;
}

const defaultPreferences = {
  itemsPerPage: 25,
};

/**
 * Renders a list of eCR data with viewer.
 * @param props - The properties passed to the component.
 * @param props.totalCount - Total number of eCRs
 * @param props.children - Contents of the wrapper
 * @returns The JSX element (table) representing the rendered list of eCRs.
 */
const EcrPaginationWrapper = ({
  totalCount,
  children,
}: EcrPaginationWrapperProps) => {
  const searchParams = useSearchParams();
  const router = useRouter();
  const pathname = usePathname();

  const currentPage = Number(searchParams.get("page")) || 1;
  const [userPreferences, setUserPreferences] =
    useState<UserPreferences>(defaultPreferences);

  useEffect(() => {
    const userPreferencesString = localStorage.getItem("userPreferences");
    if (userPreferencesString) {
      const storedPrefs = JSON.parse(userPreferencesString);
      setUserPreferences({ ...defaultPreferences, ...storedPrefs });
    }
  }, []);

  useEffect(() => {
    const current = new URLSearchParams(Array.from(searchParams.entries()));
    current.set("itemsPerPage", userPreferences.itemsPerPage.toString());
    const search = current.toString();
    const query = search ? `?${search}` : "";
    router.push(`${pathname}${query}`);
  }, [userPreferences]);

  const totalPages =
    totalCount > 0 ? Math.ceil(totalCount / userPreferences.itemsPerPage) : 1;

  const startIndex =
    totalCount > 0 ? (currentPage - 1) * userPreferences.itemsPerPage + 1 : 0;

  const endIndex = Math.min(
    currentPage * userPreferences.itemsPerPage,
    totalCount,
  );

  return (
    <div className="main-container height-ecr-library flex-column flex-align-center">
      {children}
      <div className="pagination-bar width-full padding-x-3 padding-y-105 flex-align-self-stretch display-flex flex-align-center">
        <div className="flex-1">
          Showing {startIndex}-{endIndex} of {totalCount} eCRs
        </div>
        <Pagination
          currentPage={currentPage}
          totalPages={totalPages}
          maxSlots={6}
          pathname=""
          className="flex-1"
        />
        <div className="display-flex flex-align-center flex-1 flex-justify-end">
          <Label
            htmlFor="input-select"
            className="margin-top-0 margin-right-1025"
          >
            eCRs per page
          </Label>
          <Select
            id="input-select"
            name="input-select"
            value={userPreferences.itemsPerPage}
            className="styled width-11075 margin-top-0"
            onChange={(e) => {
              const updatedUserPreferences: UserPreferences = {
                ...userPreferences,
                itemsPerPage: +e.target.value,
              };
              setUserPreferences(updatedUserPreferences);
              localStorage.setItem(
                "userPreferences",
                JSON.stringify(updatedUserPreferences),
              );
            }}
          >
            <React.Fragment>
              <option value="25">25</option>
              <option value="50">50</option>
              <option value="75">75</option>
              <option value="100">100</option>
            </React.Fragment>
          </Select>
        </div>
      </div>
    </div>
  );
};

export default EcrPaginationWrapper;
