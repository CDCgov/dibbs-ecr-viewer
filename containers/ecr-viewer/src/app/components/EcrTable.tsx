import React from "react";
import { listEcrData } from "@/app/services/listEcrDataService";
import { EcrTableClient } from "@/app/components/EcrTableClient";
import {
  EcrTableNoData,
  TableHeader,
} from "@/app/components/EcrTableClientBase";
import { DateRangePeriod } from "@/app/utils/date-utils";

/**
 * eCR Table
 * @param props - The properties passed to the component.
 * @param props.currentPage - The current page to be displayed
 * @param props.itemsPerPage - The number of items to be displayed in the table
 * @param props.sortColumn - The column to sort by
 * @param props.sortDirection - The direction to sort by
 * @param props.filterDates - The date range used to filter data
 * @param props.searchTerm - The search term used to list data
 * @param props.filterConditions - (Optional) The reportable condition(s) used to filter the data
 * @param props.headers - Table headers descriptions
 * @returns - eCR Table element
 */
const EcrTable = async ({
  currentPage,
  itemsPerPage,
  sortColumn,
  sortDirection,
  searchTerm,
  filterConditions,
  filterDates,
  headers,
}: {
  currentPage: number;
  itemsPerPage: number;
  sortColumn: string;
  sortDirection: string;
  filterDates: DateRangePeriod;
  searchTerm?: string;
  filterConditions?: string[];
  headers: TableHeader[];
}) => {
  const startIndex = (currentPage - 1) * itemsPerPage;

  const data = await listEcrData(
    startIndex,
    itemsPerPage,
    sortColumn,
    sortDirection,
    filterDates,
    searchTerm,
    filterConditions,
  );

  if (data.length === 0) {
    return <EcrTableNoData headers={headers} />;
  }

  return <EcrTableClient headers={headers} data={data} />;
};

export default EcrTable;
