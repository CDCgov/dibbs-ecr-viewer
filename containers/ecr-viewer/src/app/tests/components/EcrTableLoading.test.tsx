import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import router from "next-router-mock";

import { EcrTableLoading } from "@/app/components/EcrTableLoading";

jest.mock("../../services/listEcrDataService");

jest.mock("next/navigation", () => ({
  useRouter: () => router,
  usePathname: () => router.pathname,
  useSearchParams: () => new URLSearchParams(router.asPath.split("?")[1] || ""),
}));

describe("EcrTableLoading", () => {
  describe("Snapshot test", () => {
    let container: HTMLElement;

    beforeAll(() => {
      const mockIntersectionObserver = jest.fn();
      mockIntersectionObserver.mockReturnValue({
        observe: () => null,
        unobserve: () => null,
        disconnect: () => null,
      });
      window.IntersectionObserver = mockIntersectionObserver;

      container = render(<EcrTableLoading />).container;
    });
    it("should match snapshot", () => {
      expect(container).toMatchSnapshot();
    });
    it("should pass accessibility test", async () => {
      expect(await axe(container)).toHaveNoViolations();
    });
  });
});
