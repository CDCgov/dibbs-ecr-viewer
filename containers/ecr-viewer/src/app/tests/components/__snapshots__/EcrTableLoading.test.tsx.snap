// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcrTableLoading Snapshot test should match snapshot 1`] = `
<div>
  <tbody>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
        <br />
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
      <td>
        <div
          class="grid-row"
        >
          <div
            class="loading-blob grid-col-8 loading-blob-dark-gray width-full"
          >
             
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</div>
`;
